apps/balancer-tools/src/app/apr/(utils)/calculateApr.ts

```diff
@@ -133,95 +105,67 @@ export async function calculateAPRForDateRange(
-  const extendedFrom = initialRangeInDays < 2 ? from - SECONDS_IN_DAY : from;
-
-  const res = await fetchPoolSnapshots({
-    to,
-    from: extendedFrom,
-    network,
-    poolId,
-  });
-
-  if (res.poolSnapshots.length === 0) {
-    return [0, 0];
-  }
-
-  const sortedSnapshots: PoolSnapshot[] = res.poolSnapshots.sort(
-    (a, b) => a.timestamp - b.timestamp,
-  );
-
-  if (
-    sortedSnapshots.length <= 1 &&
-    sortedSnapshots[sortedSnapshots.length - 1].timestamp <= to
-  ) {
-    const currentData = await pools.gql(network).Pool({ poolId });
-    if (currentData.pool) {
-      sortedSnapshots.push({
-        timestamp: dateToEpoch(new Date()),
-        liquidity: currentData.pool?.totalLiquidity,
-        swapVolume: currentData.pool?.totalSwapVolume,
-        pool: {
-          symbol: currentData.pool?.symbol,
-          tokens: currentData.pool?.tokens,
-        },
-        swapFees: currentData.pool?.totalSwapFee,
-      });
-      sortedSnapshots.sort((a, b) => a.timestamp - b.timestamp);
-    }
-  }
-
-  const startRoundData = sortedSnapshots.reduce((acc, snapshot) => {
-    if (snapshot.timestamp < acc.timestamp) {
-      return snapshot;
-    }
-    return acc;
-  });
-
-  const endRoundData = sortedSnapshots.reduce((acc, snapshot) => {
-    if (snapshot.timestamp > acc.timestamp) {
-      return snapshot;
-    }
-    return acc;
-  });
-
-  if (!startRoundData?.swapFees || !endRoundData?.swapFees) {
-    throw new Error(
-      `No data found for feeAPR calculation, poolId: ${poolId}, network: ${network}, from: ${from}, to: ${to}`,
-    );
-  }
-
-  const feeDiff =
-    parseInt(endRoundData?.swapFees) - parseInt(startRoundData?.swapFees);
-
-  const feeApr = 10_000 * (feeDiff / parseInt(endRoundData?.liquidity));
-  // reference for 10_000 https://github.com/balancer/balancer-sdk/blob/f4879f06289c6f5f9766ead1835f4f4b096ed7dd/balancer-js/src/modules/pools/apr/apr.ts#L85
-  const annualizedFeeApr =
-    feeApr *
-    (SECONDS_IN_YEAR / (endRoundData?.timestamp - startRoundData?.timestamp));
-
-  return [isNaN(annualizedFeeApr) ? 0 : annualizedFeeApr / 100, feeDiff];
+) {
+  const result = await db.execute(sql`
+  WITH SnapshotDiffs AS (
+    SELECT
+      FIRST_VALUE(ps.swap_fees) OVER w AS start_swap_fees,
+      LAST_VALUE(ps.swap_fees) OVER w AS end_swap_fees,
+      LAST_VALUE(ps.liquidity) OVER w AS end_liquidity,
+      LAST_VALUE(p.protocol_swap_fee_cache) OVER w AS protocol_swap_fee_cache,
+      MIN(ps.timestamp) OVER w AS start_timestamp,
+      MAX(ps.timestamp) OVER w AS end_timestamp
+    FROM
+      pool_snapshots ps
+      JOIN pools p ON p.external_id = ps.pool_external_id
+    WHERE
+      ps.pool_external_id = '${sql.raw(poolId)}'
+      AND ps.timestamp BETWEEN '${sql.raw(
+        epochToDate(from).toISOString(),
+      )}' AND '${sql.raw(epochToDate(to).toISOString())}'
+    WINDOW w AS (ORDER BY ps.timestamp RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
```

Comment:
😅 I know I'm responsible for this monster

Reply:
we worked on this over the weekend, no need to this monster anymore ;)